Rename non-type template parameter/argument to "constant" (#7587)
cplusplus/draft: source/templates.tex

and template
\grammarterm{template-argument}{s}
are treated as types for descriptive purposes, the terms

> \term{constant parameter}

Reviewer: We seem to call these "constant template parameter" and "constant template argument" throughout, so we should introduce that very term.

Author: Hum. that footnote is in a bit of a weird state... should we get rid of it now? (the paper does that otherwise)

Author: P2841 has the following

```
A template parameter has one of the following forms:
• A type template parameter is a template parameter introduced by a type-parameter.
• A constant template parameter is a template parameter introduced by a parameter-declaration.
• A type template template parameter is a template parameter introduced by a type-tt-parameter.
• A variable template template parameter is a template parameter introduced by a variable-tt-parameter.
• A concept template parameter is a template parameter introduced by a concept-tt-parameter.

Type template template parameters, variable template template parameters, and concept
template parameters are collectively referred to as template template parameters.
```

That surgery seems best left to the paper (as it modifies the grammar - by splitting type-parameter and type-tt-parameter.

We could just remove the footnote by 
> A constant template parameter is a template parameter introduced by a parameter-declaration

and add the rest of the paragraph later if you think it would produce a more consistent intermediate state.

Reviewer: I would prefer to keep the changes here as non-inventive as possible, and leave the rest to the paper.

Author: Agreed

and
\term{constant template argument}

> are used to refer to constant, non-template parameters and arguments.

Reviewer: I think I want to keep "non-type" here.

and template
\grammarterm{template-argument}{s}
are treated as types for descriptive purposes, the terms

> \term{constant template parameter}

Author: @hubert-reinterpretcast It gets struck by the paper (which updates the surrounded text), hopefully we can live with that imperfection for a few weeks.  I don't have any objection to do it now but we agreed to keep the changes minimal

> \term{constant template parameter}

Reviewer 2: > hopefully we can live with that imperfection for a few weeks

I would rather not risk needing the paper to keep things in shape.

> \term{constant template parameter}

Reviewer: It is not clear that the concept template parameters paper makes it to the Hagenberg plenary, given that we also have contracts and reflection on the plate, which take priority.  Given Hubert's view, I think I want the bulleted list from the paper (which replaces the footnote) here, because we otherwise have no definition for the term "constant template parameter", it seems.

That's still editorial, I think; we don't change the meaning of the specification.

> \term{constant template parameter}

Author: The bulleted list refers to grammar production that don't exist in the current wording

> \term{constant template parameter}

Author: I'd rather go with Hubert suggestion and remove the first part of the sentence. The fact that we have a definition in a footnote was pre-existing. I'll do that

\end{example}

\pnum
A constant template argument is non-specialized if it is the name of a constant
template parameter.

> All other constant arguments are specialized.

Reviewer 2: ```suggestion
All other constant template arguments are specialized.
```

\item \grammarterm{parameter-declaration} in a \grammarterm{lambda-declarator}
or \grammarterm{requirement-parameter-list},
unless that \grammarterm{parameter-declaration} appears in a default argument, or

> \item \grammarterm{parameter-declaration} of a (constant) \grammarterm{template-parameter}.

Reviewer 2: ```suggestion
\item \grammarterm{parameter-declaration} of a \grammarterm{template-parameter} that declares a constant template parameter.
```

Author: I considered that (and also to remove the parenthetical entirely). parameter-declaration implies a constant template parameter, I'm afraid your wording might imply the existence of some alternative

Reviewer 2: ```suggestion
\item \grammarterm{parameter-declaration} of a \grammarterm{template-parameter} (which necessarily declares a constant template parameter).
```

Reviewer 2: @cor3ntin, I think you applied the older suggestion? Jens expressed a preference for the newer one.

\item
the specified member is not a template where a template is required, or
\item

> the specified member is not a constant where a constant is required.

Reviewer 2: The sentence applies to the following (which EDG and MSVC seems to get wrong: https://godbolt.org/z/j1hxG4f8e):
```cpp
template <typename T>
auto *f() {
  static constexpr auto x = sizeof (T::M); // sizeof _unary-expression_
  return &x;
}
struct Q { using M = int; };
void g() {
  f<Q>();
}
```

```suggestion
the specified member is not a non-type, non-template where a non-type, non-template is required.
```

Author: Good catch, I completely missed that

  // Deduction fails in each of these cases:
  f<A>(0);          // \tcode{A} does not contain a member \tcode{Y}
  f<B>(0);          // The \tcode{Y} member of \tcode{B} is not a type

>   g<C>(0);          // The \tcode{N} member of \tcode{C} is not a constant

Reviewer 2: ```suggestion
  g<C>(0);          // The \tcode{N} member of \tcode{C} is not a non-type, non-template name
```

is compared with an actual type (call it
\tcode{A}),
and an attempt is made to find template argument values (a type for a type

> parameter, a value for a constant template parameter, or a template for a
> template parameter) that will make

Reviewer 2: The inconsistent use of "template parameter" versus "parameter" following the change is very confusing.

```suggestion
template parameter, a value for a constant template parameter, or a template for a
template template parameter) that will make
```
